src/mbgl/renderer/painter_clipping.cpp

```diff
@@ -5,41 +5,44 @@
 #include <mbgl/util/clip_id.hpp>
 #include <mbgl/util/string.hpp>
 #include <mbgl/gl/debugging.hpp>
+#include <mbgl/gl/draw.hpp>
 
 namespace mbgl {
 
 
 void Painter::drawClippingMasks(PaintParameters& parameters, const std::map<UnwrappedTileID, ClipID>& stencils) {
     MBGL_DEBUG_GROUP("clipping masks");
 
-    auto& plainShader = parameters.shaders.plain;
-    auto& arrayCoveringPlain = parameters.shaders.coveringPlainArray;
-
-    mat4 matrix;
-    const GLuint mask = 0b11111111;
-
-    config.program = plainShader.getID();
-    config.stencilOp.reset();
-    config.stencilTest = GL_TRUE;
-    config.depthTest = GL_FALSE;
-    config.depthMask = GL_FALSE;
-    config.colorMask = { GL_FALSE, GL_FALSE, GL_FALSE, GL_FALSE };
-    config.stencilMask = mask;
-
-    arrayCoveringPlain.bind(plainShader, tileStencilBuffer, BUFFER_OFFSET_0, store, config);
-
     for (const auto& stencil : stencils) {
-        const auto& id = stencil.first;
-        const auto& clip = stencil.second;
+        const UnwrappedTileID& id = stencil.first;
+        const ClipID& clip = stencil.second;
+        const auto& program = parameters.shaders.plain;
 
         MBGL_DEBUG_GROUP(std::string{ "mask: " } + util::toString(id));
+
+        mat4 matrix;
         state.matrixFor(matrix, id);
         matrix::multiply(matrix, projMatrix, matrix);
-        plainShader.u_matrix = matrix;
 
-        const GLint ref = (GLint)(clip.reference.to_ulong());
-        config.stencilFunc = { GL_ALWAYS, ref, mask };
-        MBGL_CHECK_ERROR(glDrawArrays(GL_TRIANGLES, 0, (GLsizei)tileStencilBuffer.index()));
+        gl::draw(program,
+                 gl::Parameters {
+                     gl::Depth::disabled(),
+                     gl::Stencil {
+                         gl::Stencil::Always(),
+                         GLint(clip.reference.to_ulong()),
+                         0b11111111,
+                         gl::Stencil::Keep,
+                         gl::Stencil::Keep,
+                         gl::Stencil::Replace
+                     },
+                     gl::Blend::disabled(),
+                     gl::ColorMask::disabled()
+                 },
+                 gl::AttributeSources {},
+                 gl::Uniforms {
+                     gl::UniformMatrix<4> { program.u_matrix, matrix }
+                 },
```

Comment:
Yes, I think we can accomplish this with a higher-level API that layers on top of this one.